### PR TITLE
feat: flush memtable HTTP API

### LINKED ISSUE #27: add method to flush memtable manually
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Add method to flush memtable manually

**Description**
Currently memtable flush is controlled by [should_flush_table](https://github.com/ceresdb/CeresDB/blob/33e95a42277fced611ca50d39e4f3541d1555430/analytic_engine/src/table/data.rs#L381), it would be convenient to provide a method to flush memtable explicitly, to release memory pressure or to persist data to underlying objectstore 

**Proposal**
add a HTTP endpoint to flush memtable

### PR DESCRIPTION
**Pull request by jiacai2050**
# Which issue does this PR close?

Closes #27  
# Rationale for this change
 

See issue above.

# What changes are included in this PR?

 Add a new HTTP endpoint
```
POST /flush_memtable
```

# Are there any user-facing changes?

No
# How does this change test

Test manually.

This is what data directory look like after write two rows
```
$ tree         
.               
├── manifest
│   ├── 000003.log
│   ├── CURRENT        
│   ├── IDENTITY      
│   ├── LOCK
│   ├── LOG            
│   ├── MANIFEST-000001
│   └── OPTIONS-000005                                                            
├── store
└── wal
    ├── 000003.log
    ├── CURRENT
    ├── IDENTITY
    ├── LOCK
    ├── LOG
    ├── MANIFEST-000001
    └── OPTIONS-000005
```
This is what data directory look like after flush
```
$ tree
.
├── manifest
│   ├── 000003.log
│   ├── CURRENT
│   ├── IDENTITY
│   ├── LOCK
│   ├── LOG
│   ├── MANIFEST-000001
│   └── OPTIONS-000005
├── store
│   ├── 1
│   │   └── 1099511627777
│   │       └── 1.sst
│   └── 2
│       └── 2199023255553
│           └── 1.sst
└── wal
    ├── 000003.log
    ├── CURRENT
    ├── IDENTITY
    ├── LOCK
    ├── LOG
    ├── MANIFEST-000001
    └── OPTIONS-000005
```

### CODE REVIEW

File: server/src/http.rs [hunk 1]
```diff
@@ -91,19 +98,18 @@ impl<C, Q> Service<C, Q> {
     }
 }
 
-// TODO(yingwen): How to support non json response?
```
Comment (jiacai2050) [inline server/src/http.rs:94]: There is no issue return non-json reponse now, like `reply::html(...)`

File: server/src/http.rs [hunk 2]
```diff
@@ -135,6 +141,61 @@ impl<C: CatalogManager + 'static, Q: QueryExecutor + 'static> Service<C, Q> {
...
+        warp::path!("flush_memtable")
+            .and(warp::post())
+            .and(self.with_instance())
+            .and_then(|instance: InstanceRef<C, Q>| async move {
+                let get_all_tables = || {
+                    let mut tables = Vec::new();
+                    for catalog in instance
+                        .catalog_manager
+                        .all_catalogs()
+                        .map_err(|e| Box::new(e) as Box<dyn StdError + Sync + Send>)
```
Comment (waynexia) [inline server/src/http.rs:156]: maybe compilier can infer this type
```suggestion
                        .map_err(|e| Box::new(e) as _)
```

File: server/src/http.rs [hunk 3]
```diff
@@ -135,6 +141,61 @@ impl<C: CatalogManager + 'static, Q: QueryExecutor + 'static> Service<C, Q> {
...
+                        .context(Internal)?
+                    {
+                        for schema in catalog
+                            .all_schemas()
+                            .map_err(|e| Box::new(e) as Box<dyn StdError + Sync + Send>)
```
Comment (waynexia) [inline server/src/http.rs:161]:
```suggestion
                            .map_err(|e| Box::new(e) as _)
```

File: server/src/http.rs [hunk 4]
```diff
@@ -135,6 +141,61 @@ impl<C: CatalogManager + 'static, Q: QueryExecutor + 'static> Service<C, Q> {
...
+                            .context(Internal)?
+                        {
+                            for table in schema
+                                .all_tables()
+                                .map_err(|e| Box::new(e) as Box<dyn StdError + Sync + Send>)
```
Comment (waynexia) [inline server/src/http.rs:166]:
```suggestion
                                .map_err(|e| Box::new(e) as _)
```

**Review by waynexia — COMMENTED** [review]
Unrelated: I have seen this pattern (loop over catalog over schema over tables) many times, wondering if we can remove this boilerplate. But this is beyond the scope of this PR.

Other parts look good to me, cc @ShiKaiWi

File: server/src/http.rs (hunk 3 above)
Comment (jiacai2050) [reply server/src/http.rs]: fixed

**Review by waynexia — APPROVED** [review]

**Pull request merged**